### DISCUSSION TITLE
EXC_BAD_ACCESS combining OneOf, .case, .memberwise

### DISCUSSION THREAD

**Discussion by tikitu** [pointfreeco/swift-parsing#298] [Q&A] [answered]
I'm getting segmentation faults (in Xcode 14.2) and EXC_BAD_ACCESS (in Xcode 14.3) running the ParserPrinter below. I'm sorry the "minimal example" is not more minimal, but all the ways I've tried to make it smaller remove the symptom (including removing code that isn't even called). This is with version 0.12.0 of the library.

```swift
import Foundation
import Parsing

@main
struct Script {
    static func main() throws {
        let package = try setupPyFileP.parse(example)
        print(package)
    }
}

let example = """
setup(
    name='md.widget',
    version=version)  # yapf: disable
"""

let doubleQuotedStringP: some ParserPrinter<Substring, Substring>
= ParsePrint(input: Substring.self) {
    "\""
    PrefixUpTo("\"")
    "\""
}
let stringP: some ParserPrinter<Substring, Substring>
= ParsePrint {
//    OneOf {
//        singleQuotedStringP
        doubleQuotedStringP
//    }
}
struct PackageName {
    let value: Substring
}
enum DoNotCare {
    case string(Substring)
}
enum SetupArg {
    case name(PackageName)
    case doNotCare(Substring, DoNotCare)
}

// This isn't even *used* but if we remove it the parsing gets (much) further
let listWeDoNotCareP: some ParserPrinter<Substring, Substring>
= ParsePrint(input: Substring.self) {
    "["
    PrefixUpTo("]")
    "]"
}

let nameArgP: some ParserPrinter<Substring, SetupArg>
= ParsePrint(.case(SetupArg.name)) {
    "name="
    doubleQuotedStringP.map(.memberwise(PackageName.init))
}

let doNotCareArgP: some ParserPrinter<Substring, SetupArg>
= ParsePrint(.case(SetupArg.doNotCare)) {
    PrefixUpTo("=")
    "="
    doubleQuotedStringP.map(.case(DoNotCare.string))
}

let setupPyFileP: some ParserPrinter<Substring, (Substring, SetupArg, Substring)>
= ParsePrint(input: Substring.self) {
    PrefixThrough("setup(")
    OneOf {
        nameArgP
        doNotCareArgP
    }
    Rest()
}
```

**Comment by mbrandonw** [comment, marked as answer]
I can't say I know why the parser is crashing, but I do know a fix. I think Swift has a hard time dealing with large, complex types held in variables, double so for file-scope variables, and triply so when mixed with `some P` types. I converted a single one of your parsers to the more modern `body` style of parsing:

```swift
struct DoubleQuoteString: ParserPrinter {
  var body: some ParserPrinter<Substring, Substring> {
    ParsePrint {
      "\""
      PrefixUpTo("\"")
      "\""
    }
  }
}
```

And then updated all instances of `doubleQuoteString` with `DoubleQuoteString()` and it started working. Well, the parser is failing, but at least it isn't crashing.

I would recommend updating all of your parsers to this style. They will compile much faster and hopefully avoid problems like this in the future.

Also I am going to convert this to a discussion because I'm not sure there's anything we can do in the library, but it does make for a good discussion.

**Reply by tikitu** [reply]
That's absolutely brilliant, thank you so much!

**Reply by mbrandonw** [reply]
Oh also I naively converted that parser, but really you can drop the surrounding `ParsePrint` since it is implicit in the builder context:

```swift
struct DoubleQuoteString: ParserPrinter {
  var body: some ParserPrinter<Substring, Substring> {
    "\""
    PrefixUpTo("\"")
    "\""
  }
}
```

**Reply by tikitu** [reply]
Yup, these are really nice. I think I still need some `ParsePrint` for conversions (`ParsePrint(.case(MyEnum.whatever))` since you can't hang a `map` outside the body) but mostly these clean up beautifully.

**Comment by tikitu** [comment]
There's a second possible fix: take the variables out of file-scope. If I put them instead as locals inside a huge `makeParser()` function, the crash also disappears. (Best guess maybe it's caused by the lazy evaluation of global variables.)

The `body` construction with explicit types is great though, I'm all-in on that.
